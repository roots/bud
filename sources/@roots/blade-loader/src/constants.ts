export const scriptLoaders = [
  {loader: `@roots/blade-loader/script-loader`, type: `asset/source`},
  {loader: `@roots/blade-loader/asset-loader`, type: `asset/source`},
]

export const scriptPattern = /\.(js|mjs|jsx|ts|tsx|php)$/
