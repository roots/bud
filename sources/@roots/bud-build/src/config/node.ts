import type {Factory} from './index.js'

export const node: Factory<`node`> = async ({hooks}) =>
  hooks.filter(`build.node`, false)
