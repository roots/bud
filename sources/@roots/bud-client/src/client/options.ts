let options: Options = {
  timeout: 20 * 1000,
  reload: true,
  name: 'bud',
  debug: true,
  log: true,
  path: '/__bud/hmr',
  indicator: true,
  overlay: true,
}

export const set = (overrides: Options) =>
  Object.assign(options, overrides)

export const get = () => options

export const parseURLParameters = (query: string): Options =>
  Object.entries(new URLSearchParams(query)).reduce(
    (a: Options, [k, v]: [keyof Options, any]) => {
      if (v === 'true') v = true
      if (v === 'false') v = false
      return {...a, [k]: v}
    },
    options,
  )

export {options}
