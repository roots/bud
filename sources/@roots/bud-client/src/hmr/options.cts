declare global {
  const __webpack_public_path__: string
}

let options = {
  path: '/__bud/hmr',
  timeout: 20 * 1000,
  reload: false,
  name: 'bud',
}

const override = overrides => {
  if (typeof overrides.path !== 'undefined') options.path = overrides.path

  if (typeof overrides.timeout !== 'undefined')
    options.timeout = overrides.timeout

  if (typeof overrides.reload !== 'undefined')
    options.reload = overrides.reload === true

  if (typeof overrides.name !== 'undefined') options.name = overrides.name
}

export {options, override}
