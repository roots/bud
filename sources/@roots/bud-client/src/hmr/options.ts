import querystring from 'querystring'

import type {Options} from '.'

let options: Options = {
  timeout: 20 * 1000,
  reload: false,
  name: 'bud',
  path: '/__bud/hmr',
  indicator: true,
  overlay: true,
}

export const set = (overrides: Options) =>
  Object.assign(options, overrides)

export const get = () => options

export const parseQuery = (query: string): Options => {
  options = Object.entries(querystring.parse(query.slice(1))).reduce(
    (a: Options, [k, v]: [keyof Options, any]) => {
      if (v === 'true') v = true
      if (v === 'false') v = false
      return {...a, [k]: v}
    },
    options,
  )

  return options
}

export {options}
