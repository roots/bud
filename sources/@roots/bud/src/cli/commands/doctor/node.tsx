import figures from '@roots/bud-support/figures'
import {Box, Text} from '@roots/bud-support/ink'

import {LabelBox} from '../../components/LabelBox.js'

const color = {
  fail: `red`,
  success: `green`,
  warn: `yellow`,
}
const figure = {
  fail: figures.cross,
  success: figures.tick,
  warn: figures.warning,
}

const getColor = (state: `fail` | `success` | `warn`) => color[state]

const getFigure = (state: `fail` | `success` | `warn`) => figure[state]

export const Node = () => {
  let nodeState: `fail` | `success` | `warn` = `success`

  const major = Number(process.version.match(/^v(\d\d)/)?.[1] ?? 0)
  if (major < 18) nodeState = `fail`
  if (major < 20) nodeState = `warn`

  return (
    <LabelBox flexDirection="row" label="Node">
      <Box flexDirection="row" gap={1}>
        <Text color={getColor(nodeState)}>{getFigure(nodeState)}</Text>
        <Text>{process.version}</Text>
      </Box>

      {nodeState === `warn` && (
        <Text>
          Please upgrade to Node v20 for long-term support. You are running
          node {process.version}.
        </Text>
      )}

      {nodeState === `fail` && (
        <Text>
          Please upgrade to Node v20 for long-term support. You are running
          node {process.version}.
        </Text>
      )}
    </LabelBox>
  )
}
