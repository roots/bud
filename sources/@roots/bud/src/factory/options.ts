import {Config} from '@roots/bud-framework'

import {extensions} from '../extensions/index.js'
import {seed} from '../seed.js'
import {services} from '../services/index.js'

export const mergeOptions: (
  context: Config.Context,
  overrides?: Config.Options,
) => Config.Options = (context, overrides) => ({
  name: 'bud',
  mode: 'production',
  ...(overrides ?? {}),
  context: {
    ...context,
    ...(overrides?.context ?? {}),
  },
  seed: {
    ...seed,
    ...(overrides?.seed ?? {}),
  },
  services: {
    ...services,
    ...(overrides?.services ?? {}),
  },
  extensions: [...(extensions ?? []), ...(overrides?.extensions ?? [])],
})
