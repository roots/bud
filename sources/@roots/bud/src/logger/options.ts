import figures from '@roots/bud-support/figures'
import type {Config} from '@roots/bud-support/signale'

/**
 * Instance configuration
 */
export const config: Config = {
  displayScope: true,
  displayBadge: true,
  displayDate: false,
  displayFilename: false,
  displayLabel: true,
  displayTimestamp: false,
  underlineLabel: false,
  underlineMessage: false,
  underlinePrefix: false,
  underlineSuffix: false,
  uppercaseLabel: false,
}

export const level = [`error`, `warn`, `debug`, `info`]

export const types = {
  error: {
    badge: figures.cross,
    color: `red`,
    label: `error`,
    logLevel: `error`,
  },
  fatal: {
    badge: figures.cross,
    color: `red`,
    label: `fatal`,
    logLevel: `error`,
  },
  warn: {
    badge: figures.warning,
    color: `yellow`,
    label: `warning`,
    logLevel: `warn`,
  },
  success: {
    badge: figures.tick,
    color: `green`,
    label: `success`,
    logLevel: `debug`,
  },
  log: {
    badge: figures.nodejs,
    color: `blue`,
    label: `log`,
    logLevel: `debug`,
  },
  info: {
    badge: figures.info,
    color: `cyan`,
    label: `info`,
    logLevel: `info`,
  },
  debug: {
    badge: figures.circleFilled,
    color: `red`,
    label: `log`,
    logLevel: `info`,
  },
}
