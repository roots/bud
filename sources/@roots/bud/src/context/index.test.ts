import {paths} from '@repo/constants'
import {join} from 'node:path'
import {beforeAll, describe, expect, it} from 'vitest'
import {Context} from '@roots/bud-framework/options'

import getContext from './index.js'

describe(`context.get`, () => {
  let context: Context

  beforeAll(async () => {
    context = await getContext(
      join(paths.root, `tests`, `util`, `project`),
      true,
    )
  })

  it(`should be accessible`, () => {
    expect(context).toBeDefined()
  })

  it(`should match expectations`, async () => {
    expect(context.args).toEqual(
      expect.objectContaining({
        basedir: expect.any(String),
        browser: undefined,
        cache: undefined,
        ci: undefined,
        clean: undefined,
        debug: undefined,
        devtool: undefined,
        discovery: undefined,
        dry: undefined,
        editor: undefined,
        esm: undefined,
        flush: undefined,
        hash: undefined,
        html: undefined,
        immutable: undefined,
        indicator: undefined,
        input: undefined,
        level: undefined,
        log: undefined,
        manifest: undefined,
        minimize: undefined,
        mode: undefined,
        notify: true,
        output: undefined,
        overlay: undefined,
        publicPath: undefined,
        reload: undefined,
        splitChunks: undefined,
        target: undefined,
      }),
    )

    expect(context.basedir).toEqual(expect.stringMatching(/\/project$/))
    expect(context.bud).toEqual(
      expect.objectContaining({
        basedir: expect.stringMatching(/\/bud$/),
        label: `bud`,
        manifestPath: expect.stringMatching(/\/package.json$/),
        version: `0.0.0`,
      }),
    )
    expect(context.colorDepth).toEqual(256)
    expect(context.extensions.builtIn).toEqual(
      expect.arrayContaining([
        `@roots/bud-terser`,
        `@roots/bud-extensions/cdn`,
        `@roots/bud-extensions/esm`,
        `@roots/bud-extensions/fix-style-only-entrypoints`,
        `@roots/bud-extensions/clean-webpack-plugin`,
        `@roots/bud-extensions/webpack-provide-plugin`,
        `@roots/bud-extensions/webpack-manifest-plugin`,
        `@roots/bud-extensions/webpack-hot-module-replacement-plugin`,
        `@roots/bud-extensions/webpack-define-plugin`,
        `@roots/bud-extensions/mini-css-extract-plugin`,
        `@roots/bud-extensions/copy-webpack-plugin`,
        `@roots/bud-extensions/webpack-define-plugin`,
      ]),
    )
    expect(context.manifest.bud).toMatchSnapshot()
    expect(context.mode).toBe(null)
    expect(context.services).toEqual(
      expect.arrayContaining([
        `@roots/bud/services/env`,
        `@roots/bud-hooks`,
        `@roots/bud-api`,
        `@roots/bud-build`,
        `@roots/bud-cache`,
        `@roots/bud-compiler`,
        `@roots/bud-dashboard`,
        `@roots/bud-extensions`,
        `@roots/bud-server`,
        `@roots/bud/services/project`,
      ]),
    )
  })

  it(`has expected context.config`, () => {
    expect(context.config).toEqual(
      expect.objectContaining({
        '.eslintrc.js': expect.any(Object),
        'bud.config.mjs': expect.any(Object),
        'docker-compose.yml': expect.any(Object),
        'package.json': expect.any(Object),
        'tailwind.config.cjs': expect.any(Object),
        'tsconfig.json': expect.any(Object),
      }),
    )
  })
})
