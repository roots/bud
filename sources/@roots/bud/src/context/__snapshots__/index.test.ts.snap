// Vitest Snapshot v1

exports[`context.get > should match expectations 1`] = `
{
  "extensions": {
    "denylist": [
      "@roots/bud-babel",
    ],
    "discover": false,
  },
}
`;
